tools/nnyeah/nnyeah/AssemblyComparator/EventFoundEventArgs.cs

```diff
@@ -0,0 +1,22 @@
+﻿using System;
+
+#nullable enable
+
+namespace Microsoft.MaciOS.AssemblyComparator {
```

Comment:
This is a lot of boiler plate (88 lines in all, over 4 files) for type safety, and I see a potential problem:

- Since EventFoundEventArgs derives from EventNotFoundEventArgs, you could pretty easily fire the event with the wrong Args type and the type system won't complain.

Since there aren't any consumers (test or code), I'm not sure how much overkill this is vs a Found and NotFound event with a type string/enum.

Reply:
Worth considering changing the structure for the inheritance, but what you're not seeing is that the boiler plate code caught errors in the consuming code before it checked it (because it wouldn't compile).

Reply:
I do not believe that the compiler will help you at any point with the following bug:
```csharp
EventNotFound.Invoke (this, new EventFoundEventArgs ("Foo", "Bar"));
```
The above code has a bug that the compiler won't find, why? because you are inheriting from EventNotFoundEventArgs which is what @chamons  is referring too. You have to remove the inheritance from the NotFound and Found, in this case the compiler is not helping you at all. 

Regarding type safety, that is possibly an issue, but there is not need to create 1 event type per type. Create a generic event and then use the generic in the event handlers. for example:
```csharp
class ItemNotFound<T> : EventArgs {
....
}

EventHandler<ItemoNotFound<MyType>> {}
``` 
This allows several things:

1. Remove code repetition by not needing to create the 4 classes
2. Remove code repetition by not needed to do those if statements to decide how to create the event art when you fire it.
3. Have type safety.

We can leave the if statements when you fire the event if it is easier to read, but I would create a map between the type and the two events, use a generic fire that takes T as the type of the event to tise in the arguments, something like:
```csharp
FireEvent<T> (T member) {
   var arg  = new ItemFound<T> (member.Signature);
   var event = map[typeof(T)];
   event.Invoke(arg);
}
```

Reply:
I have added a sample of how I would have implemented here: https://github.com/stephen-hawley/xamarin-macios/pull/1